### PR TITLE
fix: only dial in one direction

### PR DESCRIPTION
**Pull request by dirkmc**

### CODE REVIEW

File: nodes/connect.go [hunk 1]
```diff
@@ -66,23 +69,55 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
...
 	}
 
-	connectPeers, gctx := errgroup.WithContext(ctx)
+	// Work out which addresses each node should dial, such that all dials
+	// are in one direction (nodes don't dial a node that dialed them).
+	// This helps avoid a libp2p bug where nodes that simultaneously dial
+	// each other can get random disconnects.
+	conns := make(map[peer.ID]map[peer.ID][]string)
+	for _, n := range ns {
+		npi, err := n.PeerInfo(ctx)
```
Comment (hinshun) [inline nodes/connect.go]: `n.PeerInfo` is a network call that we previously made on line 48. I think we should make `peerInfos []peerstore.PeerInfo` to a `peerInfoByNodeID map[string]peerstore.PeerInfo` so we don't have to make the call again.

File: nodes/connect.go [hunk 2]
```diff
@@ -66,23 +69,55 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
...
+		if err != nil {
+			return err
+		}
+
+		conns[npi.ID] = make(map[peer.ID][]string)
+		for _, pi := range peerInfos {
+			if _, ok := conns[pi.ID]; !ok && pi.ID != npi.ID {
```
Comment (hinshun) [inline nodes/connect.go]: To reduce nesting and improve readability we should do this:
```go
_, ok := conns[pi.ID]
if ok || pi.ID == npi.ID {
  continue
}

var peerAddrs []string
// ...
```

File: nodes/connect.go [hunk 3]
```diff
@@ -66,23 +69,55 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
...
 	for _, n := range ns {
-		n := n
-		connectPeers.Go(func() error {
-			return n.Run(gctx, metadata.Task{
-				Type:    metadata.TaskConnect,
-				Subject: strings.Join(peerAddrs, ","),
-			})
-		})
-	}
+		npi, err := n.PeerInfo(ctx)
```
Comment (hinshun) [inline nodes/connect.go]: Avoid extra network call for information we already have.

File: nodes/connect.go [hunk 4]
```diff
@@ -66,23 +69,55 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
...
+		if err != nil {
+			return err
+		}
 
-	err = connectPeers.Wait()
-	if err != nil {
-		return err
+		toConns, ok := conns[npi.ID]
+		if ok {
```
Comment (hinshun) [inline nodes/connect.go]: Prefer reducing nesting:
```go
if !ok {
  continue
}

for _, peerAddrs := range toConns {
  // ...
```

File: labapp/approuter/router.go [hunk 5]
```diff
@@ -137,6 +140,29 @@ func (s *router) connect(ctx context.Context, addrs []string) error {
...
+	span, ctx := traceutil.StartSpanFromContext(ctx, "approuter.connectOne")
+	defer span.Finish()
+	span.SetTag("addrs", len(addrs))
+
+	// Try to connect to each address in turn, until one of them works
+	var grperr error
+	for _, addr := range addrs {
+		infos, err := parseAddrs([]string{addr})
+		if err != nil {
+			grperr = err
```
Comment (hinshun) [inline labapp/approuter/router.go]: I think we should fail on `parseAddrs` errors? Are there invalid addrs that we want to ignore?

Comment (dirkmc) [reply labapp/approuter/router.go]: I believe this will fail on parseArrs errors, unless I'm reading it wrong?

Comment (hinshun) [reply labapp/approuter/router.go]: Sorry, I meant fail fast. If the next `addr` is valid, the previous `grperr` is overwritten.

Comment (dirkmc) [reply labapp/approuter/router.go]: Oh I see, that seems reasonable

Comment (dirkmc) [reply labapp/approuter/router.go]: I'll change

File: nodes/connect.go [hunk 6]
```diff
@@ -66,23 +70,62 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
...
 	}
 
-	connectPeers, gctx := errgroup.WithContext(ctx)
+	// Work out which addresses each node should dial, such that all dials
+	// are in one direction (nodes don't dial a node that dialed them).
+	// This helps avoid a libp2p bug where nodes that simultaneously dial
+	// each other can get random disconnects.
+	conns := make(map[peer.ID]map[peer.ID][]string)
+	for _, n := range ns {
+		npi, err := n.PeerInfo(ctx)
```
Comment (hinshun) [inline nodes/connect.go]: This should use `peerInfoByNodeID` too

File: nodes/connect.go [hunk 7]
```diff
@@ -39,7 +41,8 @@ func Connect(ctx context.Context, ns []p2plab.Node) error {
 	zerolog.Ctx(ctx).Info().Msg("Retrieving peer infos")
 	go logutil.Elapsed(gctx, 20*time.Second, "Retrieving peer infos")
 
-	var peerAddrs []string
+	// var peerInfos []peerstore.PeerInfo
```
Comment (hinshun) [inline nodes/connect.go]: Let's get rid of this.

**Pull request merged**